In [1]:
'''
Author: Gang Hu
Personal Email: huyingguo2007@hotmail.com
Work/School Email: ghu3@ur.rochester.edu
Date: 2023-08-16 23:54:35
LastEditTime: 2023-08-16 23:55:12
LastEditors: Gang Hu
FilePath: \Hackathon-Summer-2023\test.ipynb
'''

# Read and load the data
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scanpy as sc
import anndata as ad


dirpath = 'test_data\sce.h5ad'

# load the data
adata = sc.read_h5ad(dirpath)
len(adata) # row number


21954

Tutorial on scanpy: https://scanpy.readthedocs.io/en/stable/

In [2]:
!ls test_data

sce.h5ad
sce.rds


In [3]:
!pip install leidenalg


[notice] A new release of pip is available: 23.0 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
!pip install louvain


[notice] A new release of pip is available: 23.0 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
print(adata)

AnnData object with n_obs × n_vars = 21954 × 25269
    uns: 'X_name'


AnnData object with n_obs × n_vars = 21954 × 25269
    uns: 'X_name'

There are 21954 cells and 25269 gene features in the dataset.

In [6]:
adata.var

Empty DataFrame
Columns: []
Index: [A1BG, A1BG-AS1, A1CF, A2M, A2M-AS1, A2ML1, A4GALT, AA415398, AA465934, AA467197, AA474331, AA474408, AA987161, AAAS, AACS, AADAT, AAED1, AAGAB, AAK1, AAMDC, AAMP, AANAT, AAR2, AARD, AARS, AARS2, AARSD1, AASDH, AASDHPPT, AASS, AATF, AATK, AB015752.3, AB041803, ABAT, ABCA1, ABCA10, ABCA12, ABCA2, ABCA3, ABCA5, ABCA6, ABCA7, ABCA8, ABCA9, ABCB1, ABCB10, ABCB4, ABCB5, ABCB6, ABCB7, ABCB8, ABCB9, ABCC1, ABCC10, ABCC11, ABCC2, ABCC3, ABCC4, ABCC5, ABCC6, ABCC8, ABCC9, ABCD1, ABCD2, ABCD3, ABCD4, ABCE1, ABCF1, ABCF2, ABCF3, ABCG1, ABCG2, ABCG4, ABHD1, ABHD10, ABHD11, ABHD12, ABHD12B, ABHD13, ABHD14A, ABHD14B, ABHD15, ABHD16A, ABHD16B, ABHD17A, ABHD17B, ABHD17C, ABHD2, ABHD3, ABHD4, ABHD5, ABHD6, ABHD8, ABI1, ABI2, ABI3, ABI3BP, ABL1, ABL2, ...]

[25269 rows x 0 columns]

In [7]:
adata.obs

Empty DataFrame
Columns: []
Index: [AGAGCGAAGATCTGCT, ACATCAGTCTGACCTC, TTCTCCTGTCCTGCTT, GAAACTCTCATGTAGC, CAGTCCTGTCTAGCCG, CATGCCTCACACCGCA, AGGCCGTAGGCCCGTT, AAGTCTGAGCGTCTAT, CAGTAACCAGGAACGT, GTCTCGTTCGTACCGG, CCACGGAAGCACACAG, CGCTATCAGTGAAGTT, CGGAGCTCACATTTCT, AGATTGCAGCTGATAA, AGATTGCGTTTGGGCC, AGCGTATTCGACCAGC, GTGTTAGTCGTCGTTC, ACGGGCTCAGGTCTCG, TACTCATAGTACGTTC, CAACCTCGTACGCACC, CGATGTATCTCACATT, GATCGTAGTAAATGTG, CATGACATCTATCGCC, CATATGGCAACTGCGC, ACCGTAAGTCTAGCGC, TGGTTCCAGTCCTCCT, AGCAGCCAGTGCCATT, GGCGTGTTCAGGCAAG, CACAAACAGTGCGATG, ATGGGAGCAGATCGGA, ATCACGATCACTATTC, CTTCTCTAGCCCTAAT, AACTGGTTCGGTTCGG, GACAGAGAGCCCGAAA, ACGATGTAGCGATGAC, CCGTGGACATCTACGA, GGAACTTCATAGGATA, TCGGGACCAAAGAATC, TACACGATCGCTTAGA, ATGTGTGTCCAAAGTC, GCTGCGATCTTCGGTC, GCAATCACAGCTCGAC, GCAGTTACACATGGGA, ATCATCTTCTGCTGCT, GCTGGGTAGACTCGGA, TTCTACACACCGTTGG, CTCGTACCAATGGAGC, CGAGAAGCATGCGCAC, TGAGGGACAGCCTGTG, GGACAGATCCCAAGAT, CATTCGCGTACCAGTT, GTTTCTAAGTCGATAA, CCCATACTCGGTGTTA, ACTTTCAAGATGAGAG, CGCTATCTCCTTTACA, GGGTCTGCAGACTCGC, AAAGATGTCCGCGCAA, ACCCACTGTAAGTGTA, GTCAAGTCAATCCAAC, GGGATGAGTCGAACAG, CTGCGGAGTGTGACGA, TCGTAGAGTAAATACG, TATCTCACACGAGGTA, CGAATGTCAAGGCTCC, CACCTTGCATCACGTA, CACATTTGTATGAAAC, GTCCTCATCCTTGGTC, GGACATTCATTAGCCA, GATCTAGGTCAATACC, TGACAACGTAAGGGCT, TAGTGGTAGGCCCGTT, TGACTAGCACAGGCCT, CCCAATCCAATCGGTT, AGAATAGAGTGACTCT, ACACCAATCCGTCAAA, GTAACGTAGAGTACCG, GCGCGATCATGGAATA, AAACGGGGTGCAACTT, TGAAAGACAAGGTGTG, TGCTGCTAGGGTTTCT, GCTGCGACATGCCTAA, CGTTCTGCAATGGAGC, CCACGGACAAGTAATG, TTATGCTTCTCGGACG, GGAATAATCCATTCTA, TTGAACGCAATTGCTG, TGGCTGGCAGGACGTA, TGGTTCCTCGCCTGAG, GGATGTTCAAGAAGAG, GAACGGATCACAGGCC, CTGCGGATCTTGTACT, TTGACTTGTAACGTTC, TTTACTGGTTGATTGC, AACTCAGGTAGGACAC, AGGCCGTTCACATACG, GGACGTCGTACACCGC, AGCGGTCGTACGACCC, CTCGTCACAGGACGTA, GAGGTGAGTAATCACC, GGGCACTGTCTTTCAT, ...]

[21954 rows x 0 columns]

In [8]:
adata.X


<21954x25269 sparse matrix of type '<class 'numpy.float64'>'
	with 11039225 stored elements in Compressed Sparse Column format>

In [9]:
print(adata.uns['X_name'])


counts


In [10]:
print(adata.uns.keys())


dict_keys(['X_name'])


In [11]:
print(adata.shape)

(21954, 25269)


In [12]:
print(adata.filename)

None


In [13]:
print(adata.var[:20])

Empty DataFrame
Columns: []
Index: [A1BG, A1BG-AS1, A1CF, A2M, A2M-AS1, A2ML1, A4GALT, AA415398, AA465934, AA467197, AA474331, AA474408, AA987161, AAAS, AACS, AADAT, AAED1, AAGAB, AAK1, AAMDC]


In [14]:
df = pd.DataFrame(adata.X)
print(df.head())


                                                   0
0    (0, 14)\t1.0\n  (0, 16)\t2.0\n  (0, 17)\t2.0...
1    (0, 4)\t1.0\n  (0, 13)\t1.0\n  (0, 16)\t5.0\...
2    (0, 4)\t1.0\n  (0, 13)\t3.0\n  (0, 16)\t6.0\...
3    (0, 13)\t1.0\n  (0, 15)\t1.0\n  (0, 17)\t1.0...
4    (0, 13)\t3.0\n  (0, 15)\t1.0\n  (0, 16)\t3.0...


In [15]:
expression_matrix = adata.X.toarray() if hasattr(adata.X, 'toarray') else adata.X

df = pd.DataFrame(expression_matrix, index=adata.obs.index, columns=adata.var.index)

print(df)

                  A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  AA415398  \
AGAGCGAAGATCTGCT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ACATCAGTCTGACCTC   0.0       0.0   0.0  0.0      1.0    0.0     0.0       0.0   
TTCTCCTGTCCTGCTT   0.0       0.0   0.0  0.0      1.0    0.0     0.0       0.0   
GAAACTCTCATGTAGC   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CAGTCCTGTCTAGCCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
...                ...       ...   ...  ...      ...    ...     ...       ...   
TTTATGCTCTTGTACT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ATAACGCCAAGCTGGA   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ATTACTCAGTGGACGT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CGTGTCTCAAGCGTAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CCGGGATCAATCGGTT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   

                  AA465934 

In [16]:
df

A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  AA415398  \
AGAGCGAAGATCTGCT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ACATCAGTCTGACCTC   0.0       0.0   0.0  0.0      1.0    0.0     0.0       0.0   
TTCTCCTGTCCTGCTT   0.0       0.0   0.0  0.0      1.0    0.0     0.0       0.0   
GAAACTCTCATGTAGC   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CAGTCCTGTCTAGCCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
...                ...       ...   ...  ...      ...    ...     ...       ...   
TTTATGCTCTTGTACT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ATAACGCCAAGCTGGA   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ATTACTCAGTGGACGT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CGTGTCTCAAGCGTAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CCGGGATCAATCGGTT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   

                  AA465934  AA467197  ...  hsa-mir-3181  hsa-mir-335  \
AGAGCGAAGATCTGCT       0.0       0.0  ...           0.0          0.0   
ACATCAGTCTGACCTC       0.0       0.0  ...           0.0          0.0   
TTCTCCTGTCCTGCTT       0.0       0.0  ...           0.0          0.0   
GAAACTCTCATGTAGC       0.0       0.0  ...           0.0          0.0   
CAGTCCTGTCTAGCCG       0.0       0.0  ...           0.0          0.0   
...                    ...       ...  ...           ...          ...   
TTTATGCTCTTGTACT       0.0       0.0  ...           0.0          0.0   
ATAACGCCAAGCTGGA       0.0       0.0  ...           0.0          0.0   
ATTACTCAGTGGACGT       0.0       0.0  ...           0.0          0.0   
CGTGTCTCAAGCGTAG       0.0       0.0  ...           0.0          0.0   
CCGGGATCAATCGGTT       0.0       0.0  ...           0.0          0.0   

                  hsa-mir-4259  hsa-mir-4537  hsa-mir-4538  hsa-mir-490  \
AGAGCGAAGATCTGCT           0.0           0.0           0.0          0.0   
ACATCAGTCTGACCTC           0.0           0.0           0.0          0.0   
TTCTCCTGTCCTGCTT           0.0           0.0           0.0          0.0   
GAAACTCTCATGTAGC           0.0           0.0           0.0          0.0   
CAGTCCTGTCTAGCCG           0.0           0.0           0.0          0.0   
...                        ...           ...           ...          ...   
TTTATGCTCTTGTACT           0.0           0.0           0.0          0.0   
ATAACGCCAAGCTGGA           0.0           0.0           0.0          0.0   
ATTACTCAGTGGACGT           0.0           0.0           0.0          0.0   
CGTGTCTCAAGCGTAG           0.0           0.0           0.0          0.0   
CCGGGATCAATCGGTT           0.0           0.0           0.0          0.0   

                  hsa-mir-5195  hsa-mir-6080  hsa-mir-8072  snoU2-30  
AGAGCGAAGATCTGCT           0.0           0.0           2.0       0.0  
ACATCAGTCTGACCTC           0.0           0.0           1.0       0.0  
TTCTCCTGTCCTGCTT           0.0           0.0           1.0       0.0  
GAAACTCTCATGTAGC           0.0           0.0           2.0       0.0  
CAGTCCTGTCTAGCCG           0.0           0.0           0.0       0.0  
...                        ...           ...           ...       ...  
TTTATGCTCTTGTACT           0.0           0.0           0.0       0.0  
ATAACGCCAAGCTGGA           0.0           0.0           0.0       0.0  
ATTACTCAGTGGACGT           0.0           0.0           0.0       0.0  
CGTGTCTCAAGCGTAG           0.0           0.0           0.0       0.0  
CCGGGATCAATCGGTT           0.0           0.0           0.0       0.0  

[21954 rows x 25269 columns]

In [ ]:
training = pd.read_csv('train/training.csv', names=['UMI', 'Binary'])
training

UMI  Binary
0    GCTCTGTCAATGGATA       1
1    GATGAGGGTACGAAAT       0
2    AGGCCACGTACAGCAG       0
3    ACGCCGAGTCACACGC       0
4    TCTTCGGAGGCTAGCA       0
..                ...     ...
995  ACTGCTCCACTCGACG       0
996  TGGTTAGGTAAACGCG       0
997  TTCTCAATCAGTACGT       1
998  TGAGCCGGTCTCTTAT       0
999  AGGTCCGTCCATGCTC       0

[1000 rows x 2 columns]

In [ ]:
df

A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  AA415398  \
AGAGCGAAGATCTGCT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ACATCAGTCTGACCTC   0.0       0.0   0.0  0.0      1.0    0.0     0.0       0.0   
TTCTCCTGTCCTGCTT   0.0       0.0   0.0  0.0      1.0    0.0     0.0       0.0   
GAAACTCTCATGTAGC   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CAGTCCTGTCTAGCCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
...                ...       ...   ...  ...      ...    ...     ...       ...   
TTTATGCTCTTGTACT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ATAACGCCAAGCTGGA   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
ATTACTCAGTGGACGT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CGTGTCTCAAGCGTAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   
CCGGGATCAATCGGTT   0.0       0.0   0.0  0.0      0.0    0.0     0.0       0.0   

                  AA465934  AA467197  ...  hsa-mir-3181  hsa-mir-335  \
AGAGCGAAGATCTGCT       0.0       0.0  ...           0.0          0.0   
ACATCAGTCTGACCTC       0.0       0.0  ...           0.0          0.0   
TTCTCCTGTCCTGCTT       0.0       0.0  ...           0.0          0.0   
GAAACTCTCATGTAGC       0.0       0.0  ...           0.0          0.0   
CAGTCCTGTCTAGCCG       0.0       0.0  ...           0.0          0.0   
...                    ...       ...  ...           ...          ...   
TTTATGCTCTTGTACT       0.0       0.0  ...           0.0          0.0   
ATAACGCCAAGCTGGA       0.0       0.0  ...           0.0          0.0   
ATTACTCAGTGGACGT       0.0       0.0  ...           0.0          0.0   
CGTGTCTCAAGCGTAG       0.0       0.0  ...           0.0          0.0   
CCGGGATCAATCGGTT       0.0       0.0  ...           0.0          0.0   

                  hsa-mir-4259  hsa-mir-4537  hsa-mir-4538  hsa-mir-490  \
AGAGCGAAGATCTGCT           0.0           0.0           0.0          0.0   
ACATCAGTCTGACCTC           0.0           0.0           0.0          0.0   
TTCTCCTGTCCTGCTT           0.0           0.0           0.0          0.0   
GAAACTCTCATGTAGC           0.0           0.0           0.0          0.0   
CAGTCCTGTCTAGCCG           0.0           0.0           0.0          0.0   
...                        ...           ...           ...          ...   
TTTATGCTCTTGTACT           0.0           0.0           0.0          0.0   
ATAACGCCAAGCTGGA           0.0           0.0           0.0          0.0   
ATTACTCAGTGGACGT           0.0           0.0           0.0          0.0   
CGTGTCTCAAGCGTAG           0.0           0.0           0.0          0.0   
CCGGGATCAATCGGTT           0.0           0.0           0.0          0.0   

                  hsa-mir-5195  hsa-mir-6080  hsa-mir-8072  snoU2-30  
AGAGCGAAGATCTGCT           0.0           0.0           2.0       0.0  
ACATCAGTCTGACCTC           0.0           0.0           1.0       0.0  
TTCTCCTGTCCTGCTT           0.0           0.0           1.0       0.0  
GAAACTCTCATGTAGC           0.0           0.0           2.0       0.0  
CAGTCCTGTCTAGCCG           0.0           0.0           0.0       0.0  
...                        ...           ...           ...       ...  
TTTATGCTCTTGTACT           0.0           0.0           0.0       0.0  
ATAACGCCAAGCTGGA           0.0           0.0           0.0       0.0  
ATTACTCAGTGGACGT           0.0           0.0           0.0       0.0  
CGTGTCTCAAGCGTAG           0.0           0.0           0.0       0.0  
CCGGGATCAATCGGTT           0.0           0.0           0.0       0.0  

[21954 rows x 25269 columns]

In [ ]:
df.reset_index(inplace=True)
df


index  A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  \
0      AGAGCGAAGATCTGCT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
1      ACATCAGTCTGACCTC   0.0       0.0   0.0  0.0      1.0    0.0     0.0   
2      TTCTCCTGTCCTGCTT   0.0       0.0   0.0  0.0      1.0    0.0     0.0   
3      GAAACTCTCATGTAGC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
4      CAGTCCTGTCTAGCCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
...                 ...   ...       ...   ...  ...      ...    ...     ...   
21949  TTTATGCTCTTGTACT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21950  ATAACGCCAAGCTGGA   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21951  ATTACTCAGTGGACGT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21952  CGTGTCTCAAGCGTAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21953  CCGGGATCAATCGGTT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   

       AA415398  AA465934  ...  hsa-mir-3181  hsa-mir-335  hsa-mir-4259  \
0           0.0       0.0  ...           0.0          0.0           0.0   
1           0.0       0.0  ...           0.0          0.0           0.0   
2           0.0       0.0  ...           0.0          0.0           0.0   
3           0.0       0.0  ...           0.0          0.0           0.0   
4           0.0       0.0  ...           0.0          0.0           0.0   
...         ...       ...  ...           ...          ...           ...   
21949       0.0       0.0  ...           0.0          0.0           0.0   
21950       0.0       0.0  ...           0.0          0.0           0.0   
21951       0.0       0.0  ...           0.0          0.0           0.0   
21952       0.0       0.0  ...           0.0          0.0           0.0   
21953       0.0       0.0  ...           0.0          0.0           0.0   

       hsa-mir-4537  hsa-mir-4538  hsa-mir-490  hsa-mir-5195  hsa-mir-6080  \
0               0.0           0.0          0.0           0.0           0.0   
1               0.0           0.0          0.0           0.0           0.0   
2               0.0           0.0          0.0           0.0           0.0   
3               0.0           0.0          0.0           0.0           0.0   
4               0.0           0.0          0.0           0.0           0.0   
...             ...           ...          ...           ...           ...   
21949           0.0           0.0          0.0           0.0           0.0   
21950           0.0           0.0          0.0           0.0           0.0   
21951           0.0           0.0          0.0           0.0           0.0   
21952           0.0           0.0          0.0           0.0           0.0   
21953           0.0           0.0          0.0           0.0           0.0   

       hsa-mir-8072  snoU2-30  
0               2.0       0.0  
1               1.0       0.0  
2               1.0       0.0  
3               2.0       0.0  
4               0.0       0.0  
...             ...       ...  
21949           0.0       0.0  
21950           0.0       0.0  
21951           0.0       0.0  
21952           0.0       0.0  
21953           0.0       0.0  

[21954 rows x 25270 columns]

In [ ]:
df.rename(columns={'index': 'UMI'}, inplace=True)

df

UMI  A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  \
0      AGAGCGAAGATCTGCT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
1      ACATCAGTCTGACCTC   0.0       0.0   0.0  0.0      1.0    0.0     0.0   
2      TTCTCCTGTCCTGCTT   0.0       0.0   0.0  0.0      1.0    0.0     0.0   
3      GAAACTCTCATGTAGC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
4      CAGTCCTGTCTAGCCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
...                 ...   ...       ...   ...  ...      ...    ...     ...   
21949  TTTATGCTCTTGTACT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21950  ATAACGCCAAGCTGGA   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21951  ATTACTCAGTGGACGT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21952  CGTGTCTCAAGCGTAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
21953  CCGGGATCAATCGGTT   0.0       0.0   0.0  0.0      0.0    0.0     0.0   

       AA415398  AA465934  ...  hsa-mir-3181  hsa-mir-335  hsa-mir-4259  \
0           0.0       0.0  ...           0.0          0.0           0.0   
1           0.0       0.0  ...           0.0          0.0           0.0   
2           0.0       0.0  ...           0.0          0.0           0.0   
3           0.0       0.0  ...           0.0          0.0           0.0   
4           0.0       0.0  ...           0.0          0.0           0.0   
...         ...       ...  ...           ...          ...           ...   
21949       0.0       0.0  ...           0.0          0.0           0.0   
21950       0.0       0.0  ...           0.0          0.0           0.0   
21951       0.0       0.0  ...           0.0          0.0           0.0   
21952       0.0       0.0  ...           0.0          0.0           0.0   
21953       0.0       0.0  ...           0.0          0.0           0.0   

       hsa-mir-4537  hsa-mir-4538  hsa-mir-490  hsa-mir-5195  hsa-mir-6080  \
0               0.0           0.0          0.0           0.0           0.0   
1               0.0           0.0          0.0           0.0           0.0   
2               0.0           0.0          0.0           0.0           0.0   
3               0.0           0.0          0.0           0.0           0.0   
4               0.0           0.0          0.0           0.0           0.0   
...             ...           ...          ...           ...           ...   
21949           0.0           0.0          0.0           0.0           0.0   
21950           0.0           0.0          0.0           0.0           0.0   
21951           0.0           0.0          0.0           0.0           0.0   
21952           0.0           0.0          0.0           0.0           0.0   
21953           0.0           0.0          0.0           0.0           0.0   

       hsa-mir-8072  snoU2-30  
0               2.0       0.0  
1               1.0       0.0  
2               1.0       0.0  
3               2.0       0.0  
4               0.0       0.0  
...             ...       ...  
21949           0.0       0.0  
21950           0.0       0.0  
21951           0.0       0.0  
21952           0.0       0.0  
21953           0.0       0.0  

[21954 rows x 25270 columns]

In [ ]:
# training['UMI']
df['UMI']
# training_labels = df[mask]


0        AGAGCGAAGATCTGCT
1        ACATCAGTCTGACCTC
2        TTCTCCTGTCCTGCTT
3        GAAACTCTCATGTAGC
4        CAGTCCTGTCTAGCCG
               ...       
21949    TTTATGCTCTTGTACT
21950    ATAACGCCAAGCTGGA
21951    ATTACTCAGTGGACGT
21952    CGTGTCTCAAGCGTAG
21953    CCGGGATCAATCGGTT
Name: UMI, Length: 21954, dtype: object

In [ ]:
mask = df['UMI'].isin(training['UMI'])

training_match = df[mask].copy()

In [ ]:
# training_match.reset_index().drop('index', axis=1)
training_match = training_match.reset_index().drop('index', axis=1)
training_match

UMI  A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  \
0    AAGTCTGAGCGTCTAT   0.0       1.0   0.0  0.0      0.0    0.0     0.0   
1    AGATTGCGTTTGGGCC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
2    CATGACATCTATCGCC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
3    GTCCTCATCCTTGGTC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
4    AGGCCGTTCACATACG   0.0       1.0   0.0  0.0      0.0    0.0     0.0   
..                ...   ...       ...   ...  ...      ...    ...     ...   
995  GCTTGAACATGCCACG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
996  GGTGAAGCATTTCAGG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
997  TGTCCCATCCACGACG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
998  GCGCAGTGTTGTGGAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
999  CTTACCGCACCAACCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   

     AA415398  AA465934  ...  hsa-mir-3181  hsa-mir-335  hsa-mir-4259  \
0         0.0       0.0  ...           0.0          0.0           0.0   
1         0.0       0.0  ...           0.0          0.0           0.0   
2         0.0       0.0  ...           0.0          0.0           0.0   
3         0.0       0.0  ...           0.0          0.0           0.0   
4         0.0       0.0  ...           0.0          0.0           0.0   
..        ...       ...  ...           ...          ...           ...   
995       0.0       0.0  ...           0.0          0.0           0.0   
996       0.0       0.0  ...           0.0          0.0           0.0   
997       0.0       0.0  ...           0.0          0.0           0.0   
998       0.0       0.0  ...           0.0          0.0           0.0   
999       0.0       0.0  ...           0.0          0.0           0.0   

     hsa-mir-4537  hsa-mir-4538  hsa-mir-490  hsa-mir-5195  hsa-mir-6080  \
0             0.0           0.0          0.0           0.0           0.0   
1             0.0           0.0          0.0           0.0           0.0   
2             0.0           0.0          0.0           0.0           0.0   
3             0.0           0.0          0.0           0.0           0.0   
4             0.0           0.0          0.0           0.0           0.0   
..            ...           ...          ...           ...           ...   
995           0.0           0.0          0.0           0.0           0.0   
996           0.0           0.0          0.0           0.0           0.0   
997           0.0           0.0          0.0           0.0           0.0   
998           0.0           0.0          0.0           0.0           0.0   
999           0.0           0.0          0.0           0.0           0.0   

     hsa-mir-8072  snoU2-30  
0             0.0       0.0  
1             1.0       0.0  
2             0.0       0.0  
3             0.0       0.0  
4             0.0       0.0  
..            ...       ...  
995           0.0       0.0  
996           0.0       0.0  
997           0.0       0.0  
998           0.0       0.0  
999           0.0       0.0  

[1000 rows x 25270 columns]

In [238]:
training_match['Binary'] = training['Binary'].reset_index(drop=True)

training_match

UMI  A1BG  A1BG-AS1  A1CF  A2M  A2M-AS1  A2ML1  A4GALT  \
0    AAGTCTGAGCGTCTAT   0.0       1.0   0.0  0.0      0.0    0.0     0.0   
1    AGATTGCGTTTGGGCC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
2    CATGACATCTATCGCC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
3    GTCCTCATCCTTGGTC   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
4    AGGCCGTTCACATACG   0.0       1.0   0.0  0.0      0.0    0.0     0.0   
..                ...   ...       ...   ...  ...      ...    ...     ...   
995  GCTTGAACATGCCACG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
996  GGTGAAGCATTTCAGG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
997  TGTCCCATCCACGACG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
998  GCGCAGTGTTGTGGAG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   
999  CTTACCGCACCAACCG   0.0       0.0   0.0  0.0      0.0    0.0     0.0   

     AA415398  AA465934  ...  hsa-mir-335  hsa-mir-4259  hsa-mir-4537  \
0         0.0       0.0  ...          0.0           0.0           0.0   
1         0.0       0.0  ...          0.0           0.0           0.0   
2         0.0       0.0  ...          0.0           0.0           0.0   
3         0.0       0.0  ...          0.0           0.0           0.0   
4         0.0       0.0  ...          0.0           0.0           0.0   
..        ...       ...  ...          ...           ...           ...   
995       0.0       0.0  ...          0.0           0.0           0.0   
996       0.0       0.0  ...          0.0           0.0           0.0   
997       0.0       0.0  ...          0.0           0.0           0.0   
998       0.0       0.0  ...          0.0           0.0           0.0   
999       0.0       0.0  ...          0.0           0.0           0.0   

     hsa-mir-4538  hsa-mir-490  hsa-mir-5195  hsa-mir-6080  hsa-mir-8072  \
0             0.0          0.0           0.0           0.0           0.0   
1             0.0          0.0           0.0           0.0           1.0   
2             0.0          0.0           0.0           0.0           0.0   
3             0.0          0.0           0.0           0.0           0.0   
4             0.0          0.0           0.0           0.0           0.0   
..            ...          ...           ...           ...           ...   
995           0.0          0.0           0.0           0.0           0.0   
996           0.0          0.0           0.0           0.0           0.0   
997           0.0          0.0           0.0           0.0           0.0   
998           0.0          0.0           0.0           0.0           0.0   
999           0.0          0.0           0.0           0.0           0.0   

     snoU2-30  Binary  
0         0.0       1  
1         0.0       0  
2         0.0       0  
3         0.0       0  
4         0.0       0  
..        ...     ...  
995       0.0       0  
996       0.0       0  
997       0.0       1  
998       0.0       0  
999       0.0       0  

[1000 rows x 25271 columns]

In [240]:
training_match['Binary'].sum()

145

In [201]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import matthews_corrcoef, accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from 



X = training_match.drop(['Binary', 'UMI'], axis=1)
y = training_match['Binary']


In [202]:
X_temp, X_test, y_temp, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_temp, y_temp, test_size=0.3, random_state=42)


# clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf = RandomForestClassifier(random_state=42)


clf.fit(X_train, y_train)

y_val_pred = clf.predict(X_val)

accuracy = accuracy_score(y_val, y_val_pred)
mcc = matthews_corrcoef(y_val, y_val_pred)


print(f"Accuracy on validation set: {accuracy*100:.2f}%")
print(f"Matthews Correlation Coefficient (MCC) on validation set: {mcc:.3f}")

Accuracy on validation set: 85.24%
Matthews Correlation Coefficient (MCC) on validation set: 0.045


In [203]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)



d:\Python 3.11.0\Lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [204]:
predictions = logmodel.predict(X_val)

In [206]:
print('Confusion Matrix: ', confusion_matrix(y_val, predictions))
print('MCC: ', matthews_corrcoef(y_val, predictions))

Confusion Matrix:  [[166  15]
 [ 26   3]]
MCC:  0.025357042713313262


In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim


class FeedForwardNN(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedForwardNN, self).__init__()

        # 定义第一个线性层
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        
        # 定义激活函数
        self.relu = nn.ReLU()

        # 定义第二个线性层
        self.fc2 = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        # 通过第一个线性层传递输入
        out = self.fc1(x)
        
        # 使用激活函数
        out = self.relu(out)

        # 通过第二个线性层传递输出
        out = self.fc2(out)

        return out
        

In [ ]:
model = FeedForwardNN(input_dim=784, hidden_dim=500, output_dim=10)
criterion = nn.CrossEntropyLoss()  # 假设您进行多类分类
optimizer = optim.SGD(model.parameters(), lr=0.01)

In [ ]:
for epoch in range(num_epochs):
    for data, labels in train_loader:  # 假设train_loader是您的数据加载器

        # 清除梯度
        optimizer.zero_grad()

        # 前向传递
        outputs = model(data)

        # 计算损失
        loss = criterion(outputs, labels)

        # 反向传播和优化
        loss.backward()
        optimizer.step()

In [ ]:
model.eval()
with torch.no_grad():
    correct = 0
    total = 0
    for data, labels in test_loader:  # 假设test_loader是您的数据加载器
        outputs = model(data)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
        
print(f'Accuracy of the model: {100 * correct / total}%')

In [245]:
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim

# 设定随机种子以确保可重复性
torch.manual_seed(42)

# 模拟数据
n_samples = 1000
n_features = 10

data = np.random.randn(n_samples, n_features)
labels = (np.sum(data, axis=1) > 0).astype(int)  # 为了简单起见，我们用和大于0来标签
df = pd.DataFrame(data)

# 将数据转换为torch张量
X = torch.FloatTensor(data)
y = torch.LongTensor(labels)

# 定义一个简单的神经网络
class SimpleClassifier(nn.Module):
    def __init__(self, input_dim):
        super(SimpleClassifier, self).__init__()
        self.layer = nn.Linear(input_dim, 1)  # 2个输出，因为是二分类

    def forward(self, x):
        return self.layer(x)

# 初始化和训练模型
model = SimpleClassifier(n_features)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

for epoch in range(500):
    optimizer.zero_grad()
    outputs = model(X)
    loss = criterion(outputs, y)
    loss.backward()
    optimizer.step()

# 提取权重用于特征选择
weights = model.layer.weight[1] - model.layer.weight[0]
weights_abs = torch.abs(weights)

# 选择权重最大的特征
selected_feature_idx = torch.argsort(weights_abs, descending=True)
print("按重要性排序的特征索引:", selected_feature_idx.numpy())

IndexError: Target 1 is out of bounds.